### ISSUE TITLE
`np.arange` calculates range in lower precision input rather than output dtype

### ISSUE THREAD

**Issue by elizamanelli** [numpy/numpy#18625]
When using pytorch tensors and numpy in combination np.arange surprisingly misconverts tensors which leads to hard to debug errors.

### Reproducing code example:

```python
import numpy as np
import torch as torch

# no tensors, expected behaviour
expected = np.arange(2000,2002.4,0.001)

# manual tensor conversion, expected behaviour
np.arange(float(torch.tensor(2000)), float(torch.tensor(2002.4)), 0.001)

# tensor input, unexpected behaviour
unexpected = np.arange(torch.tensor(2000),torch.tensor(2002.4),0.001)

print(expected)
print(unexpected)
```
array([2000.   , 2000.001, 2000.002, ..., 2002.398, 2002.399, 2002.4  ])
array([2000.        , 2000.00097656, 2000.00195312, ..., 2002.34179688,
       2002.34277344, 2002.34375   ])

### Error message:

There is no error message.

### NumPy/Python version information:

Numpy version: 1.19.1 3.8.5
Torch version: 1.6.0
System: Ubuntu GCC 7.3.0

**Comment by seberg** [comment]
The NumPy version of this is:
```
np.arange(np.float32(2000), np.float32(2002.4), 0.001)
```
Which calculates the range in float32 instead of float64 (actually just with `obj2 - obj1` de-facto in Python).  The unexpected thing here is that you are getting float64 output instead of float32 which mismatches the precision of that calculation.  Either, the result should be float32, or the calculation should probably be done in float64 precision.

Torch tensors are float32 by default, because in deep learning you can get away with very low precision numbers but want all the speed you can get.

BUT:  That will not actually clear up all the confusion, it will just _slightly_ reduce it, so I am not even sure we should consider it a bug.  The input is a float32 and:
```
>>> float(torch.tensor(2002.4))
2002.4000244140625
```
so calculating the "range" of the arange in float64 will probably safe much less of that precision loss.  I expect it just happens to be enough in this example.

The short answer is probably: Please don't use arange for floats if you can avoid it. I realize `linspace` is annoying with the step-number, but arange is notorious about floating point precision problems.

**Comment by elizamanelli** [comment]
Thank you for the clarification! Makes sense, using torch.arange or linspace is a good workaround for now.
A warning in the future would be nice though, those errors are hard to debug.

**Comment by InessaPawson** [comment]
@elizamanelli We have reviewed the documentation related to `np.arange` and `linspace` and added a how-to-guide here: [numpy.org/doc/stable/user/how-to-partition.html](https://numpy.org/doc/stable/user/how-to-partition.html). 
If you have additional questions, don’t hesitate to reach out. 
And if we addressed all your questions, please close this issue.

**Comment by InessaPawson** [comment]
@elizamanelli I'm going to close this issue as it seems for now you don't have any additional questions on this topic.
